File: prism-android-wallet/app/src/main/java/io/iohk/atala/prism/app/views/fragments/ContactsFragment.java
package io.iohk.atala.prism.app.views.fragments;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;

import javax.inject.Inject;

import io.iohk.atala.prism.app.neo.common.OnSelectItem;
import io.iohk.cvp.R;
import io.iohk.atala.prism.app.data.local.db.model.Contact;
import io.iohk.cvp.databinding.FragmentContactsBinding;
import io.iohk.atala.prism.app.utils.ActivitiesRequestCodes;
import io.iohk.atala.prism.app.utils.IntentDataConstants;
import io.iohk.atala.prism.app.viewmodel.ContactsViewModel;
import io.iohk.atala.prism.app.viewmodel.ContactsViewModelFactory;
import io.iohk.atala.prism.app.views.fragments.utils.ActionBarUtils;
import io.iohk.atala.prism.app.views.fragments.utils.AppBarConfigurator;
import io.iohk.atala.prism.app.views.fragments.utils.RootAppBar;
import io.iohk.atala.prism.app.views.utils.adapters.ContactsRecyclerViewAdapter;

public class ContactsFragment extends CvpFragment<ContactsViewModel> implements OnSelectItem<Contact> {

    @Inject
    ContactsViewModelFactory factory;

    private ContactsRecyclerViewAdapter adapter;

    private FragmentContactsBinding binding;

    private ContactsViewModel viewModel;


    @Override
    public ContactsViewModel getViewModel() {
        ContactsViewModel viewModel = ViewModelProviders.of(this, factory)
                .get(ContactsViewModel.class);
        return viewModel;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        ActionBarUtils.setupMenu(menu);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        viewModel = getViewModel();
        binding = DataBindingUtil.inflate(inflater, getViewId(), container, false);
        binding.setLifecycleOwner(this);
        binding.setViewModel(viewModel);
        binding.setScanQrBtnOnClick(v -> scanQr());
        configureRecyclerView();
        initObservers();
        return binding.getRoot();
    }

    private void configureRecyclerView() {
        adapter = new ContactsRecyclerViewAdapter(this);
        binding.contactsRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        binding.contactsRecyclerView.setAdapter(adapter);
    }

    private void initObservers() {
        viewModel.getContacts().observe(getViewLifecycleOwner(), contacts -> {
            adapter.clear();
            adapter.addAll(contacts);
            adapter.notifyDataSetChanged();
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // TODO This has to be redone implementing the navigation library
        ActionBarUtils.menuItemClicked(navigator, item, this);
        // If we got here, the user's action was not recognized.
        // Invoke the superclass to handle it.
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected int getViewId() {
        return R.layout.fragment_contacts;
    }

    @Override
    protected AppBarConfigurator getAppBarConfigurator() {
        return new RootAppBar(R.string.contacts);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == ActivitiesRequestCodes.QR_SCANNER_REQUEST_ACTIVITY) {
            if (resultCode == Activity.RESULT_OK && data.hasExtra(IntentDataConstants.QR_RESULT)) {
                final String token = data.getStringExtra(IntentDataConstants.QR_RESULT);
                // TODO momentary solution, the use of "safeArgs" will be implemented
                AcceptConnectionDialogFragment dialog = AcceptConnectionDialogFragment.Companion.build(token);
                dialog.show(requireActivity().getSupportFragmentManager(), null);
            }
            return;
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onSelect(Contact item) {
        ContactDetailFragment fragment = ContactDetailFragment.Companion.build(item.id);
        navigator.showFragmentOnTop(requireActivity().getSupportFragmentManager(), fragment);
    }
}

File: prism-android-wallet/app/src/main/java/io/iohk/atala/prism/app/ui/main/contacts/ContactsFragment.kt
package io.iohk.atala.prism.app.ui.main.contacts

import android.Manifest
import android.app.Activity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.setFragmentResultListener
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import dagger.android.support.DaggerFragment
import io.iohk.atala.prism.app.data.local.db.model.Contact
import io.iohk.atala.prism.app.neo.common.IntentUtils
import io.iohk.atala.prism.app.neo.common.OnSelectItem
import io.iohk.atala.prism.app.neo.common.extensions.buildActivityResultLauncher
import io.iohk.atala.prism.app.neo.common.extensions.buildRequestPermissionLauncher
import io.iohk.atala.prism.app.ui.commondialogs.AddQrCodeDialogFragment
import io.iohk.atala.prism.app.ui.utils.adapters.ContactsRecyclerViewAdapter
import io.iohk.atala.prism.app.utils.IntentDataConstants
import io.iohk.atala.prism.app.utils.PermissionUtils
import io.iohk.cvp.BuildConfig
import io.iohk.cvp.R
import io.iohk.cvp.databinding.FragmentContactsBinding
import javax.inject.Inject

class ContactsFragment : DaggerFragment(), OnSelectItem<Contact> {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    private val viewModel: ContactsViewModel by lazy {
        ViewModelProviders.of(this, viewModelFactory).get(ContactsViewModel::class.java)
    }

    // Launcher for QrCodeScannerActivity
    private val qrActivityResultLauncher = buildActivityResultLauncher { activityResult ->
        if (activityResult.resultCode == Activity.RESULT_OK && activityResult.data?.hasExtra(IntentDataConstants.QR_RESULT) == true) {
            val token = activityResult.data!!.getStringExtra(IntentDataConstants.QR_RESULT)!!
            navigateToAcceptConnectionDialog(token)
        }
    }

    private val cameraPermissionLauncher = buildRequestPermissionLauncher { permissionGranted ->
        if (permissionGranted) showQRScanner()
    }

    private lateinit var binding: FragmentContactsBinding

    private val adapter = ContactsRecyclerViewAdapter(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setFragmentResultListener(AddQrCodeDialogFragment.KEY_REQUEST_ADD_QR) { _, bundle ->
            if (bundle.containsKey(AddQrCodeDialogFragment.KEY_RESULT_CODE)) {
                val token = bundle.getString(AddQrCodeDialogFragment.KEY_RESULT_CODE)!!
                navigateToAcceptConnectionDialog(token)
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        setHasOptionsMenu(true)
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_contacts, container, false)
        binding.lifecycleOwner = this
        binding.viewModel = viewModel
        binding.setScanQrBtnOnClick {
            showQRScanner()
        }
        configureRecyclerView()
        initObservers()
        return binding.root
    }

    private fun configureRecyclerView() {
        binding.contactsRecyclerView.layoutManager = LinearLayoutManager(context)
        binding.contactsRecyclerView.adapter = adapter
    }

    private fun initObservers() {
        viewModel.contacts.observe(
            viewLifecycleOwner,
            { contacts ->
                adapter.clear()
                adapter.addAll(contacts)
                adapter.notifyDataSetChanged()
            }
        )
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) = inflater.inflate(R.menu.contacts_menu, menu)

    override fun onPrepareOptionsMenu(menu: Menu) {
        if (BuildConfig.DEBUG) menu.findItem(R.id.action_debug_add_new_connection)?.isVisible = true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_new_connection -> {
                showQRScanner()
                true
            }
            R.id.action_debug_add_new_connection -> {
                findNavController().navigate(R.id.action_contactsFragment_to_addQrCodeDialogFragment)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onSelect(item: Contact) {
        val direction = ContactsFragmentDirections.actionContactsFragmentToContactDetailNavigation(item.id)
        findNavController().navigate(direction)
    }

    private fun showQRScanner() {
        if (PermissionUtils.checkIfAlreadyHavePermission(requireContext(), Manifest.permission.CAMERA)) {
            val intent = IntentUtils.intentQRCodeScanner(requireContext())
            qrActivityResultLauncher.launch(intent)
        } else {
            cameraPermissionLauncher.launch(Manifest.permission.CAMERA)
        }
    }

    private fun navigateToAcceptConnectionDialog(token: String) {
        val direction = ContactsFragmentDirections.actionContactsFragmentToAcceptConnectionDialogFragment(token)
        findNavController().navigate(direction)
    }
}
